File: reference-implementations/python/test_oh_my_loop.py
"""Trust and regression tests for the Python reference implementation."""
from __future__ import annotations

import json
import tempfile
import unittest
from datetime import datetime, timedelta, timezone
from pathlib import Path

from oh_my_loop import (
    AutonomyMode,
    DecisionPattern,
    Evidence,
    FeedbackStore,
    JSONMemoryStore,
    LoopConfig,
    LoopContract,
    LoopKernel,
    LoopStatus,
    MemorySensitivity,
    MemoryStatus,
    ModelConfigurationError,
    ModelDecisionError,
    PlanExecutePattern,
    ProposedAction,
    ReactPattern,
    ReflexionPattern,
    RiskLevel,
    RiskProfile,
    SelfRefinePattern,
    StepOutcome,
    StepProposal,
    contract_for,
    route,
    verify_before_claim,
)


class RouterTests(unittest.TestCase):
    @staticmethod
    def model_decision(**overrides):
        value = {
            "schema_version": "2",
            "decision": "pattern",
            "pattern": "decision",
            "domain": "semantic-domain-from-model",
            "intent": "model-inferred intent",
            "decision_owner": "user",
            "confidence": 0.9,
            "reason": "model-inferred reason",
            "warnings": [],
            "risk": {
                "level": "low",
                "autonomy": "advisory_only",
                "reasons": [],
                "requires_human": False,
                "allow_external_action": False,
            },
            "loop": {
                "name": "model-authored-loop",
                "patterns": ["decision", "custom_observation"],
                "strategy": "Act on one hypothesis, observe, and adapt the next step.",
                "feedback_type": "mixed",
                "steps": ["Run the smallest reversible probe"],
                "adaptation_rules": ["Revise the next step when evidence contradicts the hypothesis"],
                "success_evidence": ["Fresh evidence satisfies the model-defined outcome"],
                "stop_conditions": ["Evidence passes or the loop stops making progress"],
                "max_iterations": 5,
            },
        }
        risk = dict(value["risk"])
        risk.update(overrides.pop("risk", {}))
        value.update(overrides)
        value["risk"] = risk
        return value

    @staticmethod
    def classifier(value):
        return lambda _task, _prompt: value

    def test_missing_model_fails_closed(self):
        with self.assertRaises(ModelConfigurationError):
            route("opaque input")

    def test_model_receives_semantic_protocol(self):
        observed = {}

        def classify(task, prompt):
            observed.update(task=task, prompt=prompt)
            return self.model_decision(
                decision="execute_verify",
                pattern=None,
                decision_owner="agent",
                risk={"autonomy": "auto", "allow_external_action": True},
            )

        result = route("opaque input", classify_fn=classify)
        self.assertEqual(result.decision.value, "execute_verify")
        self.assertEqual(observed["task"], "opaque input")
        self.assertIn("Never\nclassify with keyword", observed["prompt"])

    def test_model_can_compose_and_invent_loop_primitives(self):
        raw = self.model_decision(pattern="task_specific_strategy")
        result = route("opaque input", classify_fn=self.classifier(raw))
        self.assertEqual(result.pattern, "task_specific_strategy")
        self.assertEqual(result.patterns, ("decision", "custom_observation"))
        self.assertEqual(result.loop.max_iterations, 5)

    def test_critical_model_classification_is_forced_to_escalation(self):
        raw = self.model_decision(
            decision="direct_answer",
            pattern=None,
            decision_owner="agent",
            risk={"level": "critical", "autonomy": "auto", "allow_external_action": True},
        )
        result = route("opaque input", classify_fn=self.classifier(raw))
        self.assertEqual(result.decision.value, "escalate")
        self.assertEqual(result.pattern, "crisis_support")
        self.assertEqual(result.risk.level, RiskLevel.CRITICAL)
        self.assertEqual(result.decision_owner.value, "expert")
        self.assertFalse(result.risk.allow_external_action)

    def test_advisory_model_classification_cannot_execute(self):
        raw = self.model_decision(decision_owner="agent")
        result = route("opaque input", classify_fn=self.classifier(raw))
        self.assertEqual(result.decision.value, "assist_only")
        self.assertEqual(result.decision_owner.value, "user")
        self.assertFalse(result.risk.allow_external_action)

    def test_confirmation_model_classification_requires_human(self):
        raw = self.model_decision(
            decision="do_once",
            pattern=None,
            risk={"level": "high", "autonomy": "confirm_before_action", "allow_external_action": True},
        )
        result = route("opaque input", classify_fn=self.classifier(raw))
        self.assertEqual(result.decision.value, "human_confirm")
        self.assertTrue(result.risk.requires_human)

    def test_low_confidence_reduces_autonomy(self):
        raw = self.model_decision(
            decision="do_once",
            pattern=None,
            confidence=0.3,
            decision_owner="agent",
            risk={"autonomy": "auto", "allow_external_action": True},
        )
        result = route("opaque input", classify_fn=self.classifier(raw))
        self.assertEqual(result.decision.value, "assist_only")
        self.assertIn("Low model confidence", result.warnings[0])

    def test_invalid_model_output_fails_closed(self):
        with self.assertRaises(ModelDecisionError):
            route("opaque input", classify_fn=self.classifier({"schema_version": "2"}))

    def test_model_failure_fails_closed(self):
        def fail(*_):
            raise RuntimeError("offline")

        with self.assertRaises(ModelDecisionError):
            route("opaque input", classify_fn=fail)

    def test_contract_uses_model_ownership_and_default_memory(self):
        raw = self.model_decision(pattern="habit")
        routed = route("opaque input", classify_fn=self.classifier(raw))
        contract = contract_for("opaque input", routed)
        self.assertEqual(contract.decision_owner.value, "user")
        self.assertTrue(contract.memory_policy.enabled)
        self.assertTrue(contract.memory_policy.capture_candidates)
        self.assertFalse(contract.consent_to_store_memory)


class VerificationTests(unittest.TestCase):
    def test_verify_pass(self):
        result = verify_before_claim("tests pass", lambda: "0 failures", {"zero": lambda x: "0" in x})
        self.assertTrue(result.can_claim)

    def test_verify_fail(self):
        result = verify_before_claim("tests pass", lambda: "2 failures", {"zero": lambda x: "0" in x})
        self.assertFalse(result.can_claim)

    def test_empty_checks_fail_closed(self):
        self.assertFalse(verify_before_claim("anything", lambda: "output", {}).can_claim)

    def test_verify_exception_fails_closed(self):
        def boom():
            raise RuntimeError("tool failed")

        result = verify_before_claim("anything", boom, {"check": lambda _: True})
        self.assertFalse(result.can_claim)
        self.assertIn("verification_run_error", result.checks)


class PatternTests(unittest.TestCase):
    def test_react_structured_completion_returns_action_output(self):
        result = ReactPattern().run(
            "task",
            think_fn=lambda *_: "next",
            act_fn=lambda *_: {"status": "completed", "output": "answer"},
        )
        self.assertTrue(result.success)
        self.assertEqual(result.output, "answer")

    def test_react_not_done_is_not_completion(self):
        result = ReactPattern(LoopConfig(react_max_iterations=1)).run(
            "task",
            think_fn=lambda *_: "I'm not done",
            act_fn=lambda *_: {"progress_delta": 1},
        )
        self.assertFalse(result.success)

    def test_react_stagnation_stops(self):
        result = ReactPattern(LoopConfig(react_max_iterations=5, max_stagnant_iterations=2)).run(
            "task",
            think_fn=lambda *_: "continue",
            act_fn=lambda *_: {"progress_delta": 0},
        )
        self.assertEqual(result.status, "blocked")
        self.assertEqual(result.iterations, 2)

    def test_reflexion_returns_latest_best_without_scores(self):
        counter = [0]

        def attempt(*_):
            counter[0] += 1
            return f"version-{counter[0]}"

        result = ReflexionPattern(LoopConfig(reflexion_max_attempts=3)).run(
            "task", attempt_fn=attempt, verify_fn=lambda _: {"pass": False}, reflect_fn=lambda *_: "change"
        )
        self.assertEqual(result.output, "version-3")

    def test_empty_plan_cannot_succeed(self):
        result = PlanExecutePattern(LoopConfig(plan_execute_max_replans=0)).run(
            "task", plan_fn=lambda *_: [], execute_fn=lambda _: None, verify_fn=lambda *_: {"pass": True}
        )
        self.assertFalse(result.success)

    def test_self_refine_max_is_partial_not_success(self):
        result = SelfRefinePattern(LoopConfig(self_refine_max_rounds=1)).run(
            "task", generate_fn=lambda task, **_: "draft", critique_fn=lambda *_: "still weak"
        )
        self.assertFalse(result.success)
        self.assertEqual(result.status, "partial")

    def test_pattern_exception_is_result_not_crash(self):
        result = ReactPattern().run("task", think_fn=lambda *_: 1 / 0, act_fn=lambda *_: None)
        self.assertEqual(result.status, "failed")

    def test_decision_pattern_keeps_user_ownership(self):
        result = DecisionPattern().run(
            "choose",
            clarify_fn=lambda _: {"values": ["health", "family"]},
            options_fn=lambda *_: ["A", "B"],
            evaluate_fn=lambda *_: {"tradeoffs": []},
        )
        self.assertTrue(result.success)
        self.assertEqual(result.output["decision_owner"], "user")

    def test_decision_pattern_requires_alternatives(self):
        result = DecisionPattern().run(
            "choose", clarify_fn=lambda _: {}, options_fn=lambda *_: ["A"], evaluate_fn=lambda *_: {}
        )
        self.assertFalse(result.success)


class KernelTests(unittest.TestCase):
    @staticmethod
    def proposal(external=False, reversible=True):
        return StepProposal(ProposedAction("step", external=external, reversible=reversible))

    @staticmethod
    def completed(*_):
        return StepOutcome(LoopStatus.COMPLETED, output="ok", evidence=(Evidence("test", "passed"),), progress_delta=1)

    def test_completed_with_evidence(self):
        result = LoopKernel().run(LoopContract("goal"), lambda *_: self.proposal(), self.completed)
        self.assertTrue(result.success)

    def test_completion_without_evidence_is_partial(self):
        result = LoopKernel().run(
            LoopContract("goal"),
            lambda *_: self.proposal(),
            lambda *_: StepOutcome(LoopStatus.COMPLETED, output="claimed"),
        )
        self.assertEqual(result.status, LoopStatus.PARTIAL)

    def test_irreversible_requires_confirmation(self):
        result = LoopKernel().run(
            LoopContract("goal"), lambda *_: self.proposal(reversible=False), self.completed
        )
        self.assertEqual(result.status, LoopStatus.ESCALATED)

    def test_rejected_confirmation_cancels(self):
        result = LoopKernel().run(
            LoopContract("goal"),
            lambda *_: self.proposal(reversible=False),
            self.completed,
            human_confirm_fn=lambda *_: False,
        )
        self.assertEqual(result.status, LoopStatus.CANCELLED)

    def test_advisory_mode_blocks_external_action(self):
        risk = RiskProfile(RiskLevel.HIGH, ("life",), AutonomyMode.ADVISORY_ONLY, True, False)
        result = LoopKernel().run(
            LoopContract("goal", risk=risk), lambda *_: self.proposal(external=True), self.completed
        )
        self.assertEqual(result.status, LoopStatus.ESCALATED)

    def test_advisory_mode_allows_internal_decision_support(self):
        risk = RiskProfile(RiskLevel.HIGH, ("life",), AutonomyMode.ADVISORY_ONLY, True, False)
        result = LoopKernel().run(
            LoopContract("compare options", risk=risk), lambda *_: self.proposal(), self.completed
        )
        self.assertEqual(result.status, LoopStatus.COMPLETED)

    def test_critical_never_calls_downstream_decider(self):
        called = [False]
        risk = RiskProfile(RiskLevel.CRITICAL, ("crisis",), AutonomyMode.CRISIS_SUPPORT, True, False)

        def decide(*_):
            called[0] = True
            return self.proposal()

        result = LoopKernel().run(LoopContract("goal", risk=risk), decide, self.completed)
        self.assertEqual(result.status, LoopStatus.ESCALATED)
        self.assertFalse(called[0])

    def test_critical_level_never_calls_downstream_decider_even_if_profile_is_inconsistent(self):
        called = [False]
        risk = RiskProfile(RiskLevel.CRITICAL, ("crisis",), AutonomyMode.AUTO, False, True)

        def decide(*_):
            called[0] = True
            return self.proposal()

        result = LoopKernel().run(LoopContract("goal", risk=risk), decide, self.completed)
        self.assertEqual(result.status, LoopStatus.ESCALATED)
        self.assertFalse(called[0])

    def test_high_risk_auto_contract_is_invalid(self):
        risk = RiskProfile(RiskLevel.HIGH, ("high",), AutonomyMode.AUTO, False, True)
        result = LoopKernel().run(LoopContract("goal", risk=risk), lambda *_: self.proposal(), self.completed)
        self.assertEqual(result.status, LoopStatus.FAILED)

    def test_budget_stops_before_second_iteration(self):
        contract = LoopContract("goal", token_budget=5, max_iterations=3)
        executed = [False]

        def execute(*_):
            executed[0] = True
            return StepOutcome(LoopStatus.CONTINUE, progress_delta=1)

        result = LoopKernel().run(
            contract,
            lambda *_: StepProposal(ProposedAction("step"), cost_tokens=5),
            execute,
        )
        self.assertEqual(result.status, LoopStatus.BUDGET_EXHAUSTED)
        self.assertFalse(executed[0])

    def test_stale_evidence_cannot_complete(self):
        stale = Evidence(
            "test",
            "passed yesterday",
            observed_at=datetime.now(timezone.utc) - timedelta(days=1),
        )
        result = LoopKernel().run(
            LoopContract("goal", max_evidence_age_seconds=60),
            lambda *_: self.proposal(),
            lambda *_: StepOutcome(LoopStatus.COMPLETED, evidence=(stale,), progress_delta=1),
        )
        self.assertEqual(result.status, LoopStatus.PARTIAL)

    def test_stagnation_blocks(self):
        result = LoopKernel().run(
            LoopContract("goal", max_iterations=5, max_stagnant_iterations=2),
            lambda *_: self.proposal(),
            lambda *_: StepOutcome(LoopStatus.CONTINUE, progress_delta=0),
        )
        self.assertEqual(result.status, LoopStatus.BLOCKED)

    def test_invalid_contract_fails(self):
        result = LoopKernel().run(LoopContract("", max_iterations=0), lambda *_: self.proposal(), self.completed)
        self.assertEqual(result.status, LoopStatus.FAILED)


class MemoryTests(unittest.TestCase):
    def setUp(self):
        self.temp = tempfile.TemporaryDirectory()
        self.path = str(Path(self.temp.name) / "memory.json")
        self.store = JSONMemoryStore(self.path)

    def tearDown(self):
        self.temp.cleanup()

    def test_personal_memory_requires_consent(self):
        with self.assertRaises(PermissionError):
            self.store.add(kind="preference", content="private", source="user")

    def test_memory_capability_is_enabled_by_default(self):
        entry = self.store.add(
            kind="public_fact",
            content="project uses Python",
            source="repository",
            sensitivity=MemorySensitivity.PUBLIC,
        )
        self.assertEqual(entry.status, MemoryStatus.QUARANTINED)

    def test_disabled_memory_does_not_capture(self):
        store = JSONMemoryStore(str(Path(self.temp.name) / "disabled.json"), enabled=False)
        with self.assertRaises(RuntimeError):
            store.add(
                kind="public_fact",
                content="project uses Python",
                source="repository",
                sensitivity=MemorySensitivity.PUBLIC,
            )

    def test_new_memory_is_quarantined(self):
        entry = self.store.add(kind="preference", content="likes mornings", source="user", consent=True)
        self.assertEqual(entry.status, MemoryStatus.QUARANTINED)
        self.assertEqual(self.store.retrieve(), [])

    def test_approved_memory_is_retrievable(self):
        entry = self.store.add(kind="preference", content="likes mornings", source="user", consent=True)
        self.store.approve(entry.id)
        self.assertEqual(self.store.retrieve("mornings")[0].id, entry.id)

    def test_forget_removes_memory(self):
        entry = self.store.add(kind="fact", content="public fact", source="test", sensitivity=MemorySensitivity.PUBLIC)
        self.assertTrue(self.store.forget(entry.id))
        self.assertFalse(self.store.forget(entry.id))

    def test_correction_quarantines_new_version(self):
        entry = self.store.add(kind="preference", content="old", source="user", consent=True, approve=True)
        corrected = self.store.correct(entry.id, "new", consent=True)
        self.assertEqual(corrected.version, 2)
        self.assertEqual(corrected.status, MemoryStatus.QUARANTINED)

    def test_corruption_fails_loudly(self):
        Path(self.path).write_text("not json", encoding="utf-8")
        with self.assertRaises(ValueError):
            self.store.retrieve()

    def test_feedback_requires_consent(self):
        store = FeedbackStore(str(Path(self.temp.name) / "feedback.json"))
        with self.assertRaises(PermissionError):
            store.record("private task", "decision", True, 0)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: reference-implementations/typescript/tests/test_oh_my_loop.ts
/** Tests for oh_my_loop TypeScript implementation. */

import {
  MODEL_ROUTER_PROMPT,
  ModelConfigurationError,
  ModelDecisionError,
  route,
  RouteDecision,
} from "../src/router";
import type { ModelClassifier } from "../src/router";
import { verifyBeforeClaim } from "../src/verify";
import { PlanExecutePattern, ReactPattern, ReflexionPattern, SelfRefinePattern } from "../src/patterns";
import { defaultConfig, shouldDowngradeModel, shouldHumanConfirm } from "../src/config";
import type { LoopConfig } from "../src/config";

let passed = 0;
let failed = 0;

function assert(condition: boolean, message: string): void {
  if (condition) {
    passed++;
  } else {
    failed++;
    console.error(`❌ FAIL: ${message}`);
  }
}

function decision(overrides: Record<string, unknown> = {}): Record<string, unknown> {
  const base = {
    schema_version: "2",
    decision: "pattern",
    pattern: "decision",
    domain: "semantic-domain-from-model",
    intent: "model-inferred intent",
    decision_owner: "user",
    confidence: 0.9,
    reason: "model-inferred reason",
    warnings: [],
    risk: {
      level: "low",
      autonomy: "advisory_only",
      reasons: [],
      requires_human: false,
      allow_external_action: false,
    },
    loop: {
      name: "model-authored-loop",
      patterns: ["decision", "custom_observation"],
      strategy: "Act on one hypothesis, observe, and adapt the next step.",
      feedback_type: "mixed",
      steps: ["Run the smallest reversible probe"],
      adaptation_rules: ["Revise the next step when evidence contradicts the hypothesis"],
      success_evidence: ["Fresh evidence satisfies the model-defined outcome"],
      stop_conditions: ["Evidence passes or the loop stops making progress"],
      max_iterations: 5,
    },
  };
  return { ...base, ...overrides, risk: { ...base.risk, ...((overrides.risk as object | undefined) ?? {}) } };
}

const classifier = (value: unknown): ModelClassifier => async () => value;

// Router tests use structured model fixtures; they do not encode task semantics.
async function testRouterUsesModel(): Promise<void> {
  let receivedPrompt = "";
  const result = await route("opaque input", async (_task, prompt) => {
    receivedPrompt = prompt;
    return decision({ decision: "execute_verify", pattern: null, decision_owner: "agent", risk: { autonomy: "auto", allow_external_action: true } });
  });
  assert(result.decision === RouteDecision.ExecuteVerify, "model decision is used");
  assert(receivedPrompt === MODEL_ROUTER_PROMPT && receivedPrompt.includes("Never classify by keyword"), "classifier receives semantic-routing protocol");
}

async function testRouterRequiresModel(): Promise<void> {
  let failedClosed = false;
  try {
    await route("opaque input", undefined as unknown as ModelClassifier);
  } catch (error) {
    failedClosed = error instanceof ModelConfigurationError;
  }
  assert(failedClosed, "missing model classifier fails closed");
}

async function testModelCanComposeAndInventLoopPrimitives(): Promise<void> {
  const result = await route("opaque input", classifier(decision({ pattern: "task_specific_strategy" })));
  assert(result.pattern === "task_specific_strategy", "model can invent a bounded strategy label");
  assert(result.loop?.patterns.includes("custom_observation") === true, "model can compose custom loop primitives");
}

async function testCriticalPolicyGate(): Promise<void> {
  const result = await route("opaque input", classifier(decision({
    decision: "direct_answer",
    pattern: null,
    decision_owner: "agent",
    risk: { level: "critical", autonomy: "auto", allow_external_action: true },
  })));
  assert(result.decision === RouteDecision.Escalate, "critical model classification is forced to escalation");
  assert(result.pattern === "crisis_support" && result.decision_owner === "expert", "critical route uses expert-owned crisis support");
  assert(result.risk.requires_human && !result.risk.allow_external_action, "critical route blocks external automation");
}

async function testAdvisoryPolicyGate(): Promise<void> {
  const result = await route("opaque input", classifier(decision({ decision_owner: "agent" })));
  assert(result.decision === RouteDecision.AssistOnly, "advisory classification is assistance only");
  assert(result.decision_owner === "user" && !result.risk.allow_external_action, "advisory route cannot be agent-owned or externally executed");
}

async function testConfirmationPolicyGate(): Promise<void> {
  const result = await route("opaque input", classifier(decision({
    decision: "do_once",
    pattern: null,
    risk: { level: "high", autonomy: "confirm_before_action", allow_external_action: true },
  })));
  assert(result.decision === RouteDecision.HumanConfirm, "confirmation autonomy requires human confirmation");
  assert(result.risk.requires_human && !result.risk.allow_external_action, "action remains blocked until confirmation");
}

async function testLowConfidenceReducesAutonomy(): Promise<void> {
  const result = await route("opaque input", classifier(decision({
    decision: "do_once",
    pattern: null,
    decision_owner: "agent",
    confidence: 0.3,
    risk: { autonomy: "auto", allow_external_action: true },
  })));
  assert(result.decision === RouteDecision.AssistOnly, "low confidence reduces autonomy");
  assert(result.warnings.length === 1, "low confidence records a warning");
}

async function testInvalidModelOutputFailsClosed(): Promise<void> {
  let failedClosed = false;
  try {
    await route("opaque input", classifier({ schema_version: "2" }));
  } catch (error) {
    failedClosed = error instanceof ModelDecisionError;
  }
  assert(failedClosed, "invalid model output fails closed");
}

// Verify tests
function testVerifyPass(): void {
  const result = verifyBeforeClaim(
    "tests pass",
    () => "all tests passed, 0 failures",
    {
      containsPassed: (o) => String(o).includes("passed"),
      noFailures: (o) => String(o).toLowerCase().includes("0 failures"),
    },
  );
  assert(result.canClaim === true, "verify should pass when all checks pass");
}

function testVerifyFail(): void {
  const result = verifyBeforeClaim(
    "tests pass",
    () => "2 tests failed",
    {
      noFailures: (o) => !String(o).toLowerCase().includes("fail"),
    },
  );
  assert(result.canClaim === false, "verify should fail when a check fails");
}

function testVerifyEmptyFailsClosed(): void {
  const result = verifyBeforeClaim("done", () => "anything", {});
  assert(result.canClaim === false, "empty verifier must fail closed");
}

function testVerifyExceptionFailsClosed(): void {
  const result = verifyBeforeClaim("done", () => { throw new Error("offline"); }, { exists: () => true });
  assert(result.canClaim === false, "verifier exception must fail closed");
}

// Pattern tests
function testReactTerminatesOnAnswer(): void {
  const pattern = new ReactPattern();
  const result = pattern.run("test task", {
    thinkFn: () => "I have the answer",
    actFn: () => "action result",
  });
  assert(result.success === true, "react should succeed on answer");
  assert(result.iterations === 1, "react should terminate in 1 iteration");
}

function testReactTerminatesOnMax(): void {
  let count = 0;
  const pattern = new ReactPattern({ reactMaxIterations: 3 });
  const result = pattern.run("test task", {
    thinkFn: () => `thinking ${++count}`,
    actFn: () => "action result",
  });
  assert(result.success === false, "react should fail at max");
  assert(result.iterations === 3, "react should stop at 3 iterations");
}

function testReactNotDoneDoesNotTerminate(): void {
  const pattern = new ReactPattern({ reactMaxIterations: 1 });
  const result = pattern.run("test", { thinkFn: () => "I'm not done", actFn: () => "partial" });
  assert(result.success === false, "negated done phrase must not terminate");
}

function testReflexionTerminatesOnSuccess(): void {
  const pattern = new ReflexionPattern();
  const result = pattern.run("test", {
    attemptFn: () => "solution",
    verifyFn: () => ({ pass: true }),
    reflectFn: () => "reflection",
  });
  assert(result.success === true, "reflexion should succeed");
  assert(result.attempts === 1, "reflexion should terminate in 1 attempt");
}

function testReflexionTerminatesOnMax(): void {
  let count = 0;
  const pattern = new ReflexionPattern({ reflexionMaxAttempts: 3 });
  const result = pattern.run("test", {
    attemptFn: () => `attempt ${++count}`,
    verifyFn: () => ({ pass: false }),
    reflectFn: () => "try differently",
  });
  assert(result.success === false, "reflexion should fail at max");
  assert(result.attempts === 3, "reflexion should stop at 3 attempts");
  assert(result.output === "attempt 3", "reflexion should return latest/best output");
}

function testEmptyPlanFails(): void {
  const result = new PlanExecutePattern().run("test", {
    planFn: () => [], executeFn: () => "unused", verifyFn: () => ({ pass: true }),
  });
  assert(result.success === false, "empty plan cannot succeed vacuously");
}

function testSelfRefineLimitIsPartial(): void {
  const result = new SelfRefinePattern({ selfRefineMaxRounds: 1 }).run("test", {
    generateFn: () => "draft", critiqueFn: () => "still weak",
  });
  assert(result.success === false && result.status === "partial", "max refinement is partial, not success");
}

// Config tests
function testConfigDegradation(): void {
  const config: LoopConfig = { ...defaultConfig, costBudgetTokens: 10000, cheapModelThreshold: 0.5 };
  assert(shouldDowngradeModel(config, 4000) === false, "no downgrade at 40%");
  assert(shouldDowngradeModel(config, 5000) === true, "downgrade at 50%");
}

function testConfigHumanConfirm(): void {
  const config: LoopConfig = { ...defaultConfig, humanConfirmCostThreshold: 10000 };
  assert(shouldHumanConfirm(config, 9000) === false, "no confirm below threshold");
  assert(shouldHumanConfirm(config, 10000) === true, "confirm at threshold");
}

async function main(): Promise<void> {
  await testRouterUsesModel();
  await testRouterRequiresModel();
  await testModelCanComposeAndInventLoopPrimitives();
  await testCriticalPolicyGate();
  await testAdvisoryPolicyGate();
  await testConfirmationPolicyGate();
  await testLowConfidenceReducesAutonomy();
  await testInvalidModelOutputFailsClosed();
  testVerifyPass();
  testVerifyFail();
  testVerifyEmptyFailsClosed();
  testVerifyExceptionFailsClosed();
  testReactTerminatesOnAnswer();
  testReactTerminatesOnMax();
  testReactNotDoneDoesNotTerminate();
  testReflexionTerminatesOnSuccess();
  testReflexionTerminatesOnMax();
  testEmptyPlanFails();
  testSelfRefineLimitIsPartial();
  testConfigDegradation();
  testConfigHumanConfirm();

  console.log(`\n${failed === 0 ? "✅" : "❌"} ${passed} tests passed, ${failed} failed`);
  if (failed > 0) process.exitCode = 1;
}

void main().catch((error: unknown) => {
  console.error(error);
  process.exitCode = 1;
});
